src/simd/generic.rs

```diff
@@ -150,59 +224,67 @@ pub trait SIMD<
+where
+    ScalarDType: Copy + PartialOrd + AsPrimitive<usize>,
+    SIMDVecDtype: Copy,
+    SIMDMaskDtype: Copy,
+    T: SIMDOps<ScalarDType, SIMDVecDtype, SIMDMaskDtype, LANE_SIZE> + SIMDInstructionSet,
+{
+    // Use the implementation
+}
+
+// --------------- Float Ignore NaNs
+
+/// SIMD operations for setting a SIMD vector to a scalar value (only required for floats)
+pub trait SIMDSetOps<ScalarDType, SIMDVecDtype>
+where
+    ScalarDType: FloatCore,
+{
+    unsafe fn _mm_set1(a: ScalarDType) -> SIMDVecDtype;
+}
 
+/// SIMDCore trait that ignore NaNs (for float types)
+pub trait SIMDCoreIgnoreNaN<ScalarDType, SIMDVecDtype, SIMDMaskDtype, const LANE_SIZE: usize>:
+    SIMDOps<ScalarDType, SIMDVecDtype, SIMDMaskDtype, LANE_SIZE> + SIMDSetOps<ScalarDType, SIMDVecDtype>
+where
+    ScalarDType: FloatCore + AsPrimitive<usize>,
+    SIMDVecDtype: Copy,
+    SIMDMaskDtype: Copy,
+{
     #[inline(always)]
     unsafe fn _core_argminmax(arr: &[ScalarDType]) -> (usize, ScalarDType, usize, ScalarDType) {
         assert_eq!(arr.len() % LANE_SIZE, 0);
         // Efficient calculation of argmin and argmax together
         let mut new_index = Self::INITIAL_INDEX;
-        let mut index_low = Self::INITIAL_INDEX;
-        let mut index_high = Self::INITIAL_INDEX;
-
-        let increment = Self::_mm_set1(LANE_SIZE);
 
         let mut arr_ptr = arr.as_ptr(); // Array pointer we will increment in the loop
-        let mut values_low = Self::_mm_loadu(arr_ptr);
-        let mut values_high = Self::_mm_loadu(arr_ptr);
-
-        // This is (40%-5%) slower than the loop below (depending on the data type)
-        // arr.chunks_exact(LANE_SIZE)
-        //     .into_iter()
-        //     .skip(1)
-        //     .for_each(|step| {
-        //         new_index = Self::_mm_add(new_index, increment);
-
-        //         let new_values = Self::_mm_loadu(step.as_ptr());
-
-        //         let lt_mask = Self::_mm_cmplt(new_values, values_low);
-        //         let gt_mask = Self::_mm_cmpgt(new_values, values_high);
-
-        //         index_low = Self::_mm_blendv(index_low, new_index, lt_mask);
-        //         index_high = Self::_mm_blendv(index_high, new_index, gt_mask);
-
-        //         values_low = Self::_mm_blendv(values_low, new_values, lt_mask);
-        //         values_high = Self::_mm_blendv(values_high, new_values, gt_mask);
-        //     });
+        let new_values = Self::_mm_loadu(arr_ptr);
+
+        // Update the lowest values and index
+        let mask = Self::_mm_cmplt(new_values, Self::_mm_set1(ScalarDType::infinity()));
+        let mut values_low =
+            Self::_mm_blendv(Self::_mm_set1(ScalarDType::infinity()), new_values, mask);
+        let mut index_low = Self::_mm_blendv(Self::_mm_set1(ScalarDType::zero()), new_index, mask);
+
+        // Update the highest values and index
+        let mask = Self::_mm_cmpgt(new_values, Self::_mm_set1(ScalarDType::neg_infinity()));
```

Comment:
let's avoid shadowing both of these mask values - let's call it `mask_upper` and `mask_lower` perhaps?